util/progress/printer.go

```diff
@@ -98,12 +103,35 @@ func NewPrinter(ctx context.Context, w io.Writer, out console.File, mode string,
 		}
 	}
 
+	pw := &Printer{
+		ready: make(chan struct{}),
+	}
 	go func() {
-		resumeLogs := logutil.Pause(logrus.StandardLogger())
-		// not using shared context to not disrupt display but let is finish reporting errors
-		pw.warnings, pw.err = progressui.DisplaySolveStatus(ctx, c, w, statusCh, solveStatusOpt...)
-		resumeLogs()
-		close(doneCh)
+		for {
+			pw.status = make(chan *client.SolveStatus)
+			pw.done = make(chan struct{})
+
+			pw.logMu.Lock()
+			pw.logSourceMap = map[digest.Digest]interface{}{}
+			pw.logMu.Unlock()
+
+			close(pw.ready)
+
+			resumeLogs := logutil.Pause(logrus.StandardLogger())
+			// not using shared context to not disrupt display but let is finish reporting errors
+			pw.warnings, pw.err = progressui.DisplaySolveStatus(ctx, c, w, pw.status, solveStatusOpt...)
```

Comment:
`warnings` should be appended here otherwise `Pause`+`Unpause` can discard old warnings from `Warnings()`.

Reply:
Hm, agreed, I actually did try this. Unfortunately, if you run the same build twice (using `reload`), you then get duplicate warnings. I think the right call here is to discard warnings like we currently do, but print the warnings in the monitor, so we still get them?

Reply:
SGTM. Or printing warning before starting monitor might be enough for covering the case where the monitor is enabled but the user doesn't call `reload`. As of now, no warning is printed for the user in this case.

Reply:
Hm, this is actually surprisingly difficult to refactor into a way that feels natural - ideally, we'd print the warnings after every single `reload`. But to do that, the warnings logic ends up getting propagated down into the monitor which doesn't seem right to me.

@tonistiigi @crazy-max WDYT about adding a `progressui.WithOnClose()` to be able to execute on finishing the solve display, so we could define the warnings printing logic there, instead of needing to propagate the warnings printing logic throughout? Something like 6590ececd67dc9d7fbbf04320713a7b5eca66647?

Reply:
The printer object with variadic opts and on close logic looks good.

I wonder if we could also take this to our advantage and make things better for logrus output that is not synchronized properly with the TTY progressbar like: https://github.com/docker/buildx/blob/afcaa8df5fdb9c8c908c8dae325d86a0f8f8c3ef/build/build.go#L693

The Formatter interface would help: https://pkg.go.dev/github.com/sirupsen/logrus#Formatter